### ISSUE TITLE
Unexpected token ILLEGAL on regex \.

### ISSUE THREAD

**Issue by JarLob** [sebastienros/esprima-dotnet#44]
This code (regex with \\.) triggers unexpected token: `var isHtml = /\.html$/;`

**Comment by sebastienros** [comment]
Note to my future self: It works on http://esprima.org/demo/parse.html so the fix should be easy to find by just doing a step by step debug session and find out the difference between esprima and esprima-dotnet.

**Comment by KvanTTT** [comment]
Yes, I'm also very interested in fixing of lexer errors with regex.

**Comment by JarLob** [comment]
I got it in Esprima.Sample as you can see from the [call stack.](https://github.com/sebastienros/esprima-dotnet/issues/44#issuecomment-425156368)

```cs
        [Fact]
        public void ShouldRegularExpressionGH44()
        {
            var scanner = new Scanner(@"var isHtml = /\.html$/");

            var   tokens = new List<Token>();
            Token token;

            do
            {
                scanner.ScanComments();
                token = scanner.Lex();
                tokens.Add(token);
            } while (token.Type != TokenType.EOF);
        }
```

**Comment by KvanTTT** [comment]
@sebastienros it does not work only with `Scanner`, not parser.

**Comment by olliejm** [comment]
I'm still having this issue with a regex that also contains `\.`. 

I cloned dev branch and ran the unit test posted by JarLob above, it still failed with both his regex and mine. 

In my case I'm experiencing the problem via a Jint execution of a file containing the problematic regex, using Jint 3.0.0-beta-1715, with Esprima 1.0.1258

**Comment by olliejm** [comment]
Here is my particular regex:

```js
var urlRegex = /(https?)\:\/\/[A-Za-z0-9\.\-]+(\/[A-Za-z0-9\?\&\=;\+!'\(\)\*\-\._~%]*)*/gi;
```

The error occurs on this line: https://github.com/sebastienros/esprima-dotnet/blob/dev/src/Esprima/Scanner.cs#L602 when processing the first `\` escape character in the line;

**Comment by sebastienros** [comment]
I can't repro these issues if I use the parser directly, or the ScanRegex method. I think that the issue is in the Esprima.Sample source that you all seem to be following. The parser does more than just call `Lex` and does some lookaheads that will help scan the regex. Otherwise it's trying to scan an identifier instead.

Is your intent to actually iterate over each Token of a script, like the sample is supposed to work?

**Comment by IntranetFactory** [comment]
We use 

```
var urlRegex = /(https?)\:\/\/[A-Za-z0-9\.\-]+(\/[A-Za-z0-9\?\&\=;\+!'\(\)\*\-\._~%]*)*/gi;
```

in a JInt script and that triggers the error in the Esprima dependency.

**Comment by lahma** [comment]
Tested above regex with latest Jint 3 using REPL and worked just fine. Maybe someone should post a complete sample with used library versions.

**Comment by KurtGokhan** [comment]
Here is the simplest failing program: `/\./`

This is the code from the sample project with that regex, and it is throwing the error in the title.

```
        public static void Main(string[] args)
        {
            var scanner = new Scanner(@"/\./");
            Tokenize(scanner);
        }

        private static void Tokenize(Scanner scanner)
        {
            var tokens = new List<Token>();
            Token token;

            do
            {
                scanner.ScanComments();
                token = scanner.Lex();
                tokens.Add(token);
            } while (token.Type != TokenType.EOF);

            Console.WriteLine(JsonConvert.SerializeObject(tokens, Formatting.Indented));
        }
```

I encountered this error in Jint. Does Jint use Scanner?

**Comment by adams85** [comment]
Tracked this down: the root cause of the issue is the JS syntax itself, more precisely, the `/` and `/=` tokens, which are ambiguous. When the scanner encounters one of them, it can't tell whether it's the beginning of a regexp or it's a division operator *without* knowing the context. But to know the context, you need to parse the code... For more details, see the comments of [this SO answer](https://stackoverflow.com/a/4726410/8656352) (please note that the accepted answer is wrong):

> Technically, there are a couple ambiguities that are unavoidable at the lexical level. For example, (a+b)/c vs. if (x) /foo/.exec('bar') (close-paren can precede either). Also, ++ /foo/.abc and a++ / b (plus-plus can precede either). Together with -- these are the only ones I know of.

To sum it up, you can't (reliably) use the scanner in standalone mode when the code contains regexps. Which situation is kinda sad but it looks there's no solution to this problem. What we can maybe do to mitigate it is to allow the user to provide some best effort algorithm, which would enable tokenization in some specific cases instead of failing with error. What do you think?
